File: ios/CybersourceFingerprint.m
#import "CybersourceFingerprint.h"
#import <RLTMXProfiling/TMXProfiling.h>
#import <RLTMXProfilingConnections/TMXProfilingConnections.h>
#import <React/RCTLog.h>

@implementation CybersourceFingerprint
RLTMXProfiling *RLTMXProfilingInstance;

RCT_EXPORT_MODULE()

RCT_REMAP_METHOD(config,
                  withOrgId:(NSString *)orgId withServerUrl:(NSString *)serverUrl
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)
{
    RLTMXProfilingConnections *profilingConnections = [[RLTMXProfilingConnections alloc] init];

    profilingConnections.connectionTimeout = 20;
    profilingConnections.connectionRetryCount = 2;

    RLTMXProfilingInstance = [RLTMXProfiling sharedInstance];

    @try {
        [RLTMXProfilingInstance configure: @{
            RLTMXOrgID: orgId,
            RLTMXFingerprintServer: serverUrl,
            RLTMXProfileTimeout: @20,
            RLTMXProfilingConnectionsInstance: profilingConnections,
        }];
    } @catch (NSException *exception) {
        reject(@"CybersourceFingerprint", @"ERROR", nil);
    }

    resolve(@"Config finished");
}

RCT_REMAP_METHOD(startProfiling,
                  withSessionId:(NSString *)sessionId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)
{
    RCTLogInfo(@"Profiling with sessionId: %@", sessionId);

    RLTMXProfileHandle *profileHandle = [
        RLTMXProfilingInstance
        profileDeviceUsing: @{RLTMXSessionID : sessionId}
        callbackBlock: ^(NSDictionary * _Nullable result) {
            RLTMXStatusCode statusCode = [
                [result valueForKey:RLTMXProfileStatus]
                integerValue
            ];

            resolve(@{
                @"sessionId": [result valueForKey:RLTMXSessionID],
                @"status": @(statusCode),
            });
        }
    ];
}

// Don't compile this code when we build for the old architecture.
#ifdef RCT_NEW_ARCH_ENABLED
- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params
{
    return std::make_shared<facebook::react::NativeCybersourceFingerprintSpecJSI>(params);
}
#endif

@end

File: ios/CybersourceFingerprint.swift
import RLTMXProfiling
import RLTMXProfilingConnections

@objc(CybersourceFingerprint)
class CybersourceFingerprint: NSObject {
    private var profile: RLTMXProfiling = RLTMXProfiling.sharedInstance()!

    @objc(multiply:withB:withResolver:withRejecter:)
    func multiply(a: Float, b: Float, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        resolve(a*b)
    }
    
    @objc(config:fpsUrl:resolver:rejecter:)
    func config(_ orgId: String, fingerprintServerUrl: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
//        let profile = RLTMXProfiling.sharedInstance()!
        let profilingConnections: RLTMXProfilingConnections  = RLTMXProfilingConnections.init()
        profilingConnections.connectionTimeout    = 20;
        profilingConnections.connectionRetryCount = 2;
        
        profile.configure(configData:[
           RLTMXOrgID              :orgId,
           RLTMXFingerprintServer  :fingerprintServerUrl,
           RLTMXProfilingConnectionsInstance: profilingConnections,
        ])
        resolve("PUTARIA!")
    }
    
    @objc(startProfiling:resolver:rejecter:)
    func startProfiling(_ sessionId: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        let customAttributes : [String : String] = [RLTMXSessionID: sessionId]
        let profileHandle: RLTMXProfileHandle = profile.profileDevice(profileOptions: customAttributes, callbackBlock: {(result: [AnyHashable : Any]?) -> Void in
            let results:NSDictionary! = result! as NSDictionary
            let status:RLTMXStatusCode  = RLTMXStatusCode(rawValue:(results.value(forKey: RLTMXProfileStatus) as! NSNumber).intValue)!
            let iddasessao = results.value(forKey: RLTMXSessionID) as! String
//            let localresolver =
            resolve(iddasessao)
//            localresolver(status)
//            if (status == .ok) {
//                resolve()
//            }
        })
    }
}


